File: 13.py
from utils import *

ls = []
# %%
L = open('13.in','r').readlines()
T = int(L[0].rstrip('\n'))
tt = [int(x) for x in L[1].rstrip('\n').replace('x','0').split(',')]

# %%
def one():
    wait_time = 1e9
    bus_nr = 0
    for bus_line in tt:
        new_wait = bus_line - T%bus_line if bus_line!=0 else 1e9
        if new_wait < wait_time:
            wait_time = new_wait
            bus_nr = bus_line
    return bus_nr*wait_time


def two_simplex():
    # tt value is bus line, and its index is the offset
    ##################################################################
    # model = LpProblem(name="example-problem", sense=LpMinimize)
    # x1 = LpVariable(name="x1", lowBound=1, cat=LpInteger)
    # x2 = LpVariable(name="x2", lowBound=1, cat=LpInteger)
    # x3 = LpVariable(name="x3", lowBound=1, cat=LpInteger)
    # x4 = LpVariable(name="x4", lowBound=1, cat=LpInteger)
    # model += (x1 - 17*x2 == 0)
    # model += (x1 - 13*x3 == -2)
    # model += (x1 - 19*x4 == -3)
    # model += (x1 <= lcm(17,13,19))
    # model += x1 # obj_function
    ##################################################################
    model = LpProblem(name="part2-problem", sense=LpMinimize)
    sol = LpVariable(name="sol", lowBound=1, cat=LpInteger)
    xx = []
    for i,b in enumerate(tt):
        if b>0:
            x = LpVariable(name="x"+str(b),
                           lowBound=1,
                        #    upBound=lcm(*[y for y in tt if y>0])/b,
                           cat=LpInteger)
            xx.append((b,x,i))
    for eq in xx:
        model += (sol - eq[0]*eq[1] == -eq[2])
    model += (sol <= lcm(*[y for y in tt if y>0]))
    model += (sol >= 0)
    model += sol
    
    # Solve the optimization problem
    status = model.solve(PULP_CBC_CMD(msg=0)) # Set msg to 1 to print solver steps
    # print(f"status: {model.status}, {LpStatus[model.status]}")
    # print(f"objective: {model.objective.value()}")

    # for var in model.variables():
    #     print(f"{var.name}: {var.value()}")

    # for name, constraint in model.constraints.items():
    #     print(f"{name}: {constraint.value()}")
    return int(model.objective.value())

def ok_mod(base,n,mod):
    if base<n: 
        return False,n
    else:
        return ((base+mod)%n)==0,n

def two():
    bus_offset_pairs = []
    for i,b in enumerate(tt):
        if b>0:
            bus_offset_pairs.append((b,i))
    allmods_ok = False
    base = 0
    succ_base = set()
    to_jmp = set([b for b in bus_offset_pairs])
    tmp_jmp = deepcopy(to_jmp) # Reduce n of iterations if we avoid checking mods for steps we already did.
    while not allmods_ok:
        allmods_ok = True
        for bop in to_jmp:
            res, inc_base = ok_mod(base, bop[0], bop[1])
            if res: 
                succ_base.add(inc_base)
                tmp_jmp.remove(bop)
            allmods_ok &= res
        to_jmp = deepcopy(tmp_jmp)
        if not allmods_ok:
            base+= prod(succ_base)
    return base

if __name__ == "__main__":
    print(f"{one() = }") # 2298
    print(f"{two() = }") # 783685719679632

File: 13.cpp
#include "aoc.hpp"

const string inputname = "13.in";
using bus_t = int;
using timetable_t = vector<bus_t>;
using sol_t = double;
using timestamp_t = int;
using bus_offset_pair_t = pair<int,int>;

sol_t one(const timetable_t & tt, const timestamp_t T)
{
    auto wait_time = 1e9;
    auto bus_nr = 0;
    for (auto & bus_line: tt)
    {
        auto new_wait = 1e9;
        if (bus_line!=0)
        {
            new_wait = bus_line - T%bus_line;
        }
        if (new_wait < wait_time)
        {
            wait_time = new_wait;
            bus_nr = bus_line;
        }
    }
    return bus_nr*wait_time;
}

pair<bool,timestamp_t> ok_mod(sol_t base, timestamp_t n, timestamp_t mod)
{
    if (base<n)
    {
        return make_pair(false,n);
    }
    else
    {
        return make_pair(fmod(base+mod,n)==0,n);
    }
    
}

sol_t two(const timetable_t & tt)
{
    vector<bus_offset_pair_t> bus_offset_pairs{};
    set<bus_offset_pair_t> to_jmp{};
    set<bus_offset_pair_t> tmp_jmp{};
    set<timestamp_t> succ_base{};
    for (auto [i, b]: enumerate(tt))
    {
        if(b>0)
        {
            bus_offset_pair_t pbi = make_pair(b,i);
            bus_offset_pairs.push_back(pbi);
            to_jmp.insert(pbi);
            tmp_jmp.insert(pbi);
        }
    }
    auto allmods_ok = false;
    sol_t base = 0;
    while (!allmods_ok)
    {
        allmods_ok = true;
        for (auto & bop: to_jmp)
        {
            auto [res, inc_base] = ok_mod(base, bop.first, bop.second);
            if (res)
            {
                succ_base.insert(inc_base);
                tmp_jmp.erase(bop); //erase(remove(tmp_jmp.begin(), tmp_jmp.end(), bop), tmp_jmp.end());
            }
            allmods_ok &= res;
        }
        to_jmp = tmp_jmp;
        // copy(tmp_jmp.begin(), tmp_jmp.end(), to_jmp.begin());
        if (!allmods_ok)
        {
            sol_t prod = 1;
            for (auto & x: succ_base) //13,19,23,37,41
            {
                prod*=x;
            }
            base+= prod; //0,1,2,3... ~20 -> 322, 621, 11684, 465267, 9083344
        }
    }
    return base;
}

int main()
{
    ifstream infile(inputname);
    string s;
    string chars = "\n\r";
    std::regex e ("(\\d+)|(x)");
    timestamp_t T;
    regex_token_iterator<string::iterator> rend;
    getline( infile, s );
    std::regex_token_iterator<std::string::iterator> a ( s.begin(), s.end(), e );
    T = stoi(*a);
    getline( infile, s );
    std::regex_token_iterator<std::string::iterator> b ( s.begin(), s.end(), e );
    timetable_t timetable{};
    while (b!=rend)
    {
        auto val = *b++;
        if (val.str()=="x")
        {
            timetable.push_back(0);
        }
        else
        {
            timetable.push_back(stoi(val));
        }
        
    } 

    cout << "one = " << one(timetable, T) << '\n'; // 2298
    cout << "two = " << setprecision(16) << two(timetable) << '\n'; // 783685719679632
    return 0;
}